fusedrug/utils/file_formats/fasta_efficient_db_index.py

```diff
@@ -0,0 +1,174 @@
+        Create SQLite index for fast sequence lookup.
+        """
+        # SQLite database to store file offsets
+
+        already_found = os.path.isfile(self.index_file_path)
+
+        # sqlite3 does not allow sharing connections between threads (it can result in unexpected behavior)
+        # so we use a separate connection per process+thread combination
+        thread_desc = f"{os.getpid()}@{threading.get_ident()}"
+        if thread_desc not in self.connections:
+            self.connections[thread_desc] = sqlite3.connect(self.index_file_path)
+        self.conn = self.connections[thread_desc]
+        cursor = self.conn.cursor()
+
+        if not already_found:
+            seen_so_far = 0
+
+            # Create index table
+            cursor.execute(
+                """
+                CREATE TABLE IF NOT EXISTS fasta_index (
+                    seq_id TEXT PRIMARY KEY,
+                    offset INTEGER
+                )
+            """
+            )
+
+            with open(self.fasta_path, "rb") as f:
+                mmapped_file = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
+
+                # Track file position
+                current_pos = 0
+                while current_pos < len(mmapped_file):
+                    # Find next '>' marker
+                    next_marker = mmapped_file.find(b">", current_pos)
+                    if next_marker == -1:
+                        break
+
+                    # Extract sequence ID
+                    end_id = mmapped_file.find(b"\n", next_marker)
+                    seq_id = (
+                        mmapped_file[next_marker + 1 : end_id]
+                        .decode("utf-8")
+                        .split()[0]
+                    )
+
+                    if self.entry_id_process_func is not None:
+                        seq_id = self.entry_id_process_func(seq_id)
+                        assert (
+                            seq_id != ""
+                        ), "Found an empty id! make sure that your FASTA file is fine and that the provided entry_id_process_func is fine"
+
+                    # Store offset
+                    cursor.execute(
+                        "INSERT OR IGNORE INTO fasta_index VALUES (?, ?)",
+                        (seq_id, next_marker),
+                    )
+
+                    # Move to next sequence
+                    current_pos = end_id + 1
+
+                    seen_so_far += 1
+
+                    if not (seen_so_far % 1_000_000):
+                        print(f"seen_so_far={seen_so_far}")
+
+                    if not (seen_so_far % 1000):
+                        self.conn.commit()
+
+            self.conn.commit()
+
+        # #count table size (takes too long on 300M+ database, commented out for now)
+        # res = cursor.execute('''
+        #     SELECT COUNT(*) FROM fasta_index
+        # ''')
+        # res.fetchall()
+        # print(res)
+
+    def get_sequence(self, seq_id: str) -> str:
```

Comment:
Remind me what is sequence ID? the index in fasta file?
How can we also get the description?

Reply:
the sequence ID in this case:
```>tr|A0A7C4TNA9|A0A7C4TNA9_9EURY Type-4 uracil-DNA glycosylase OS=Archaeoglobus sp. OX=1872626 GN=ENS21_01095 PE=3 SV=1
METLEDIRKEILNCKKCPLHLTKKNYVPGSGSGKSGIVFVGEAPGREEDLRGEPFVGSAG
KLLTELLSSVGLTREEVFICNVLKCRPPNNRDPLPEEVEACGSYLVRQLNALKPDLIVCL```

is `tr|A0A7C4TNA9|A0A7C4TNA9_9EURY` - it's basically everything after the initial '>' up to the first space.
The description is currently discarded. I will add support for loading it in a followup PR.